File: JS/BOJ-S3-2910-Frequency_sorting.js
'use strict';

const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : '../input.txt';
let input = fs
  .readFileSync(filePath)
  .toString()
  .trim()
  .split('\n')
  .map((v) => v.trimEnd());

let output = [];
const solution = (input) => {
  const [N, C] = input[0].split(' ').map(Number);
  const nums = input[1].split(' ').map(Number);
  const mp = {};
  const mp_first = {};

  nums.forEach((v, i) => {
    if (!Object.hasOwn(mp, v)) {
      mp[v] = 1;
      mp_first[v] = i;
    } else {
      mp[v] = mp[v] + 1;
    }
  });

  nums.sort((a, b) => {
    if (mp[a] == mp[b]) {
      return mp_first[a] < mp_first[b] ? -1 : 1;
    }
    return mp[a] < mp[b] ? 1 : -1;
  });

  output.push(...nums);
};

solution(input);
fs.writeFileSync('/dev/stdout', output.join(' ').trimEnd());
// fs.writeFileSync(
//   '../output.txt',
//   output.join(' ').trimEnd(),
//   (err) => console.error
// );

File: CPP/BOJ-S3-2910-Frequency_sorting.cpp
#include <bits/stdc++.h>
using namespace std;

int n, c, tmp;
vector<int> v;
map<int, int> mp, mp_first;

bool cmp(int a, int b)
{
  if (mp[a] == mp[b])
  {
    return mp_first[a] < mp_first[b];
  }
  return mp[a] > mp[b];
}

int main()
{
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  // freopen("input.txt", "r", stdin);
  // freopen("output.txt", "w", stdout);

  cin >> n >> c;

  for (int i = 0; i < n; ++i)
  {
    cin >> tmp;
    mp[tmp]++;
    if (mp_first[tmp] == 0)
      mp_first[tmp] = i + 1;
    v.push_back(tmp);
  }

  sort(v.begin(), v.end(), cmp);

  for (auto i : v)
  {
    cout << i << " ";
  }

  return 0;
}